### DISCUSSION TITLE
p:ajax not called on f:selectItem noSelectionOption="true"

### DISCUSSION THREAD

**Discussion by vitsin** [primefaces/community#1573] [PrimeFaces] [answered]
In case user select "-- SELECT --", than renderPages() is not invoked. Why? What I am doing wrong here?
In case of actual brand's value that coming from brandSelectConverter - renderPages() invoked OK!

```
<p:selectOneMenu id="menuBrandSelect"
	value="#{testView.selectedMenu.brand}"
	converter="#{brandSelectConverter}">
		<f:selectItem itemLabel="-- SELECT --" noSelectionOption="true"/>
		<f:selectItems value="#{testView.brands}"
			var="b"
			itemLabel="#{testView.getTitle(b)}"
			itemValue="#{b}"/>
		<p:ajax listener="#{testView.renderPages(testView.selectedMenu.brand)}"
			update="@this"/>
</p:selectOneMenu>
```

```
import jakarta.faces.view.ViewScoped;
import org.springframework.stereotype.Component;

@Component
@ViewScoped
public class TestView implements Serializable {
    private selectedMenu;
    private List<Brand> brands;

    @PostConstruct
    public void init()
    {
        selectedMenu = new MyMenu();
	brands = ... list populated from database here;
    }

    public void renderPages(Brand selectedBrand)
    {
        System.out.println("selected object brand: "+selectedBrand);
    }
}
```

appreciate the help,
--vs

P.S.
i am aware of this example: http://www.primefaces.org:8080/showcase/ui/ajax/dropdown.xhtml?jfwid=42344 but it works with Strings not custom object.

**Comment by melloware** [comment]
But No SelectOption you didn't select anything so no event will fire right?  What would "Brand SelectedBrand" be when No Select Option is picked?  Can you put together an executable example using the [PrimeFaces Test](https://github.com/primefaces/primefaces-test) project?

**Reply by vitsin** [reply]
sorry, can't wrap my head around your questions.
I'll try to clarify, so if user select from GUI --SELECT--:
```
<f:selectItem itemLabel="-- SELECT --" noSelectionOption="true"/>
```

than p;ajax listener is not fired

```
<p:ajax listener="#{testView.renderPages(testView.selectedMenu.brand)}"
```

But when any other option selected, than p:ajax works fine.
Ok, if this is by design, than how to add optional --SELECT-- to list of Objects (in my case Brand obj). thank you,

**Reply by melloware** [reply]
well i don't know if its being blocked by your Converter since its NOT a Brand or whether this is by design with NoSelectOption.  Hard to tell without a deep dive or a PrimeFaces Test reproducer.

**Reply by vitsin** [reply]
@melloware  still, why p:ajax is not getting called than on empty value ?
```
<p:ajax listener="#{testView.renderPages(testView.selectedMenu.brand)}"
			update="@this"/>
```

**Reply by melloware** [reply]
Oh I thought your reply was that you solved it.  I am still waiting for a PrimeFaces Test project....

**Comment by vitsin** [comment]
when I strip down view completely to bare minimum with p:selectOneMenu+converter in the PrimeFaces Test project, than p:ajax listener get invoked correctly even in case of f:selectItem --SELECT--. May be combination of Jakarta, Spring, Faces and PF somehow causing it, will have to investigate.

**Reply by melloware** [reply]
That is the power of stripping it down. Prove it works and now figure out what combination of yours is making it not work.

**Comment by vitsin** [comment, marked as answer]
required="true" on same element preventing <p:ajax listener="testView.renderPages"/> call in case <f:selectItem itemLabel="-- SELECT --" itemValue="#{null}" noSelectionOption="true"/>.

In other words required="true" happens before p:ajax call, as designed.
